File: solutions/1762. Buildings With an Ocean View.py
class Solution:
    def findBuildings(self, heights: List[int]) -> List[int]:
        #First attempt - Time O(N), space O(1)
        n = len(heights)
        ans = [n-1]
        max_heights = heights[-1] #space optimization, O(N) -> O(1)
        
        for i in range(n-2,-1,-1):
            if heights[i] > max_heights:
                max_heights = heights[i] 
                ans.append(i)
                
        return ans[::-1]

File: solutions/1762. Buildings With an Ocean View.cpp
class Solution {
public:
    // [4,2,3,1]
    // [4,3,3,1], ans = [3,2,0]
    vector<int> findBuildings(vector<int>& heights) {
        int n=heights.size();
        vector<int> ans;
        int maxSeen = heights[n-1];
        ans.push_back(n-1);
        for (int i = n-2; i >= 0; i--) {
            if (heights[i] > maxSeen) {
                ans.push_back(i);
                maxSeen = heights[i];
            }
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};
